src/compiler/scala/tools/nsc/reporters/Reporter.scala

```diff
@@ -47,61 +50,104 @@ abstract class Reporter extends InternalReporter {
-        case InternalReporter.WARNING => warningCount < settings.maxwarns.value
-        case _       => true
-      }
-    // work around fractured API to support `reporters.Reporter.info`, which is not forwarded
-    override protected def info0(pos: Position, msg: String, severity: Severity, force: Boolean): Unit =
-      severity match {
-        case InternalReporter.ERROR      => delegate.error(pos, msg)     // for symmetry, but error and warn are already forwarded
-        case InternalReporter.WARNING    => delegate.warning(pos, msg)
-        case _ if force => delegate.echo(pos, msg)
-        case _          => delegate.info(pos, msg, force = false)
-      }
-  }
-  // mark reporters known to respect maxerrs
-  implicit def `adapt reporter to legacy API`(reporter: InternalReporter): Reporter =
-    reporter match {
-      case _: LimitFilter => new AdaptedReporter(reporter) with LimitedReporter
-      case _              => new AdaptedReporter(reporter)
-    }
-  // whitelist reporters known to respect maxerrs; otherwise, enforce user-specified reduced limit
-  def limitedReporter(settings: Settings, reporter: Reporter): Reporter =
-    reporter match {
-      case _: ConsoleReporter | _: LimitedReporter => reporter
-      case _ if settings.maxerrs.isSetByUser && settings.maxerrs.value < settings.maxerrs.default =>
-        new LimitingReporter(settings, reporter)
-      case _ => reporter
+  /** Split a message into a prefix and an optional explanation that follows a line starting with `"----"`. */
+  private def splitting(msg: String, explaining: Boolean): String =
+    if (msg != null && msg.indexOf("\n----") > 0) {
+      val (err, exp) = msg.linesIterator.span(!_.startsWith("----"))
+      if (explaining) (err ++ exp.drop(1)).mkString("\n") else err.mkString("\n")
+    } else {
+      msg
     }
 
   /** The usual way to create the configured reporter.
-   *  Errors are reported through `settings.errorFn` and also by throwing an exception.
-   */
+    *  Errors are reported through `settings.errorFn` and also by throwing an exception.
+    */
   def apply(settings: Settings): Reporter = {
     //val loader = ScalaClassLoader(getClass.getClassLoader)  // apply does not make delegate
     val loader = new ClassLoader(getClass.getClassLoader) with ScalaClassLoader
-    val res = loader.create[InternalReporter](settings.reporter.value, settings.errorFn)(settings)
-    if (res.isInstanceOf[Reporter]) res.asInstanceOf[Reporter]
-    else res: Reporter  // adaptable
+    val res = loader.create[FilteringReporter](settings.reporter.value, settings.errorFn)(settings)
+    res
+  }
+}
+
+/** The reporter used in a Global instance.
+  *
+  * It filters messages based on
+  *   - settings.nowarn
+  *   - settings.maxerrs / settings.maxwarns
+  *   - positions (only one error at a position, no duplicate messages on a position)
+  */
+abstract class FilteringReporter extends Reporter {
+  def settings: Settings
```

Comment:
It seemed obvious that FilteringReporter should take settings as a ctor param, as all reporters are required to have that ctor sig, but InteractiveReporter has an abstract compiler from which it derives settings.

Reply:
I was literally just thinking, Nobody has to muck with settings at construction time, but:
```
  /**
   * Called before instantiating the presentation compiler's instance.
   * You should provide an implementation of this method if you need
   * to customize the `settings` used to instantiate the presentation compiler.
   * */
  protected def prepareSettings(settings: Settings): Unit = ()
```
Well, how about you take the compiler as an arg and they can do all their mucking about before they call you?

Reply:
> FilteringReporter should take settings as a ctor param

Either way is fine with me, but I don't see a reason to prefer one or the other.

Reply:
It's a clear signal or reminder to customizers that you must provide this constructor. Otherwise it's just documented somewhere, or not documented. I did spend a minute wondering if there is a reason to have a preference.

Reply:
I didn't mean to persuade you!

Reply:
I'm not sure you did :-) But as I said, either way is fine with me.